Cli exec

SanchayanMaity opened the pull request:
@harendra-kumar I did this work 3 weeks back when I last commented, and now getting back to it again after holidays. I thought before I go further I did ask you, if I am on the right track or this current work does not make much sense. Thanks for your time.

SanchayanMaity committed:
Add a gitignore file

Add a gitignore file so that stack-work does not show
up in the output of "git status".

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

harendra-kumar commented on bench-show.cabal:
```diff
@@ -123,6 +123,25 @@ library
         , statistics        >= 0.15    && < 0.16
         , vector            >= 0.10    && < 0.13
 
+executable bench-show-cli-exe
```
We can call it simply `bench-show`.

harendra-kumar commented on app/Main.hs:
```diff
@@ -0,0 +1,81 @@
+
+titleField :: Parser TitleAnnotation
+titleField = flag' TitleField
+    ( long "titlefield"
```
For CLI we can use the convention to use hyphen to separate words. So a camel case like `TitleField` becomes `--title-field` on CLI.

harendra-kumar commented on app/Main.hs:
```diff
@@ -0,0 +1,81 @@
+
+groupStyle :: Parser GroupStyle
+groupStyle = flag' Absolute
+    ( long "Absolute" <> help "Show absolute field values for all groups" )
```
On CLI usually all lower case is preferred (e.g. `absolute` instead of `Absolute`) for typing convenience.

harendra-kumar replied:
For multi-value config options an argument option instead of using individual flags is a better way to present via CLI, for example `--group-style absolute` or `--group-style=absolute`.

harendra-kumar reviewed (commented):
Looks good in general. I have added a few comments. I think you can go ahead and implement, we can nitpick later.

harendra-kumar commented:
@SanchayanMaity thanks for the work! It will be really convenient and useful once done.

harendra-kumar commented:
Once you are done I will make a release and announce the tool, as it will now become convenient to play with directly from the CLI instead of writing a program. One more issue that I see is too many dependencies which makes it take long time to build.

SanchayanMaity commented:
@harendra-kumar Thanks for the feedback. Is there a way to reduce the dependencies? Also, I see from Travis CI that the build failed for an earlier version of GHC due to lack of Semigroups. So, need to add that as well.

I got stuck a bit by the way. So for the final implementation after the parsers are written, to run execparser, I need ParserInfo. So it has to be ParserInfo Config and for that I first need a Parser Config. To put together Parser Config, I have like below

```haskell
Config <$> pVerbose <*> pOutputDir <*> pTitle
            <*> pTitleAnnotation <*> pPresentation
            <*> pEstimator <*> pThreshold <*> pDiffStrategy
```
But, the rest of the fields of Config viz. Fields, Groups and Benchmarks are not being set. How do I take care of that in the above? Guess I am missing something obvious. I changed the names a little since they conflicted with the ones from lib.

Right now I am getting an error and if I am understanding it correctly, it is the above problem. The error is, in case if I am not understanding it correctly,

```haskell
/home/sanchayan/GitSources/bench-show/app/Main.hs:91:11: error:
    • Couldn't match type ‘([String] -> [String])
                           -> [(String, Double, Double)]
                           -> [(String, FieldTick)]
                           -> (String -> Maybe (String, String))
                           -> ([(String, Int)] -> [(String, Int)])
                           -> ((SortColumn -> Either String [(String, Double)]) -> [String])
                           -> Config’
                     with ‘Config’
      Expected type: Parser Config
        Actual type: Parser
                       (([String] -> [String])
                        -> [(String, Double, Double)]
                        -> [(String, FieldTick)]
                        -> (String -> Maybe (String, String))
                        -> ([(String, Int)] -> [(String, Int)])
                        -> ((SortColumn -> Either String [(String, Double)]) -> [String])
                        -> Config)
    • In the expression:
        Config <$> pVerbose <*> pOutputDir <*> pTitle <*> pTitleAnnotation
          <*> pPresentation
          <*> pEstimator
          <*> pTheshold
          <*> pDiffStrategy
      In an equation for ‘pConfig’:
          pConfig
            = Config <$> pVerbose <*> pOutputDir <*> pTitle
                <*> pTitleAnnotation
                <*> pPresentation
                <*> pEstimator
                <*> pTheshold
                <*> pDiffStrategy
   |
91 | pConfig = Config <$> pVerbose <*> pOutputDir <*> pTitle
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

SanchayanMaity commented:
@harendra-kumar So I kind of got to the current approach now. However, that use of CliConfig feels hacky to me. I could not figure out how to handle only partial update of record while using optparse. Other than this, I think I did not grasp the complete doc clearly. Each CLI option should be optional. I have to specify all of them now at the moment for it to work. Will fix it.

harendra-kumar commented:
@SanchayanMaity all config options are optional, when none is specified `defaultConfig` can be used as it is. When the user wants to specify an option he/she can update that field in `defaultConfig` and use the updated record. The only mandatory arguments are the input/output files to be specified in report/graph operations.

Now we have to translate this same mechanism to CLI. What we can do is use `Config` data type as it is and parse the CLI options into it rather than using a separate `CliConfig`, unless there is some trouble because of doing this that I may not be aware of. 

Now, the parser has to provide values for ALL the fields for the structure that is being parsed. If a field is optional, the parser has a provision to specify the default value for the field. In the default value you can just use whatever is there in `defaultConfig` e.g. for the estimator cli option you can use `estimator defaultConfig` as the default value in the parser. With this the parser would now result in a full `Config` after the parse succeeds.

Another point is that other than the `Config` you would also need the input file and output file to produce a graph/report. So you can wrap Config inside another record and add those fields too. So it may look something like this:

```
data OpType = Report | Graph

data CLIOptions = CLIOptions 
    { cliConfig :: Config
    , opType :: OpType
    , inputFile :: FilePath
    , outputFile :: Maybe FilePath
    }
```

The `outputFile` would be mandatory when `opType` is `--graph` otherwise it would be optional.

So you would be parsing the `CLIOptions` record above. One we have this we have everything to execute the operation. Based on the `opType` we will just execute `report` or `graph` op.

SanchayanMaity commented:
If output file for graph option is compulsory, it is better to do like this I would say and that's what I have done for now. 

```
data OpType = Report | Graph FilePath
```
The implementation is done with your changes except for two problems which remain.

1. Due to Config being a product type, I have a problem which I have described [here](https://stackoverflow.com/questions/54505684/how-to-parse-a-product-type-partially). That's why I make the fields that I want to parse from Config for eg. "verbose", a member of CLIOptions again.

2. Since field names I want to parse are the same in CLIOptions and Config, I get a duplicate record field error for which I used the DuplicateRecordFields extension. However, this extension is only available in GHC 8 versions and later. I see that we mention tested with GHC 7.10 and also CI tests fail because of that extension. I assume you would want me to not use this extension and rename the fields for CLIOptions.

SanchayanMaity committed:
Add support for executable with CLI options

Add support to have a convenient executable with command
line options corresponding to the Config API to generate
graphs and reports from CSV file.

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

SanchayanMaity committed:
app: Main.hs: Cleanup the Config parser implementation

Use pure from applicative and wrap up the default options
for Config that were not specified earlier. This cleans up
the CLI parser for options we want to read for Config
while keeping the defaults as per the already specified
defaultConfig.

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

SanchayanMaity committed:
app: Main.hs: Remove ScopedTypeVariables extension

Remove the unused extension. Somehow this was missed
since hlint did not give an unused extension warning.

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

harendra-kumar commented:
Thanks @SanchayanMaity ! I am merging this PR, I have a few changes that I will push in another PR.

harendra-kumar commented:
Thanks @SanchayanMaity ! I am merging this PR, I have a few changes that I will push in another PR.
